### PR TITLE
Added Interface

### LINKED ISSUE #45: Create a new DS called `Fluid dictionary`
This should be the polymorphic brother of the standard dictionary.

This is the interface we should implement:
```C#
public interface IFluidDictionary<TKey>
{
    bool Set<T>(TKey key, T value);
    bool TryGetValue<T>(TKey key, out T value);
    T GetValueOrDefault<T>(TKey key);
    T GetRequiredValue<T>(TKey key);
    IEnumerable<TKey> Keys { get; }
}
```

### PR DESCRIPTION
**Pull request by postmeback**
## Pull Request Description 

As requested in the ticket, I have added a the interface in the requested path.

Fixes #45 

## Motivation and Context

_[Explain in details the reasoning behind these changes and provide any relevant context. This can include potential impact or any alternative solutions considered.]_

## Checklist

- [x] I have tested these changes thoroughly.
- [x] I have added/updated relevant documentation.
- [x] My code follows the project's coding guidelines.
- [x] I have performed a self-review of my changes.
- [x] My changes are backwards compatible.

### CODE REVIEW

**Comment by postmeback** [comment]
It is still work in progress

**Comment by TonyTroeff** [comment]
There are no tests covering these changes and the documentation is not understandable.

**Review by TonyTroeff — CHANGES_REQUESTED** [review]
There are no tests covering these changes and the documentation is not understandable.

**Comment by postmeback** [comment]
Okay, regarding documentation and tests, I will add later

**Comment by postmeback** [comment]
Documentation updated. Requesting for review.

**Comment by postmeback** [comment]
image


Tested successfully

File: TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs [hunk 1]
```diff
@@ -0,0 +1,78 @@
+namespace TryAtSoftware.Extensions.Collections.Tests;
+
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+public class FluidDictionaryTests
```
Comment (TonyTroeff) [inline TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs:9]: In all `TryAtSoftware.*` projects we recommend using random values as much as possible. See `TryAtSoftware.Randomizer` - it exposes helper methods for generating random `string` or `int` values.

File: TryAtSoftware.Extensions.Collections/FluidDictionary.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    /// <summary>
+    /// Sets the value of type T associated with the specified key in the dictionary.
+    /// </summary>
+    /// <typeparam name="T">The type of the value to set.</typeparam>
+    /// <param name="key">The key to set the value for.</param>
+    /// <param name="value">The value to set.</param>
+    /// <returns>True if the value is successfully set, false otherwise.</returns>
+    public bool Set<T>(TKey key, T value)
+    {
+        if (key == null)
```
Comment (TonyTroeff) [inline TryAtSoftware.Extensions.Collections/FluidDictionary.cs:25]: We recommend using `is` for null checks.

File: TryAtSoftware.Extensions.Collections/FluidDictionary.cs [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    /// Attempts to get the value associated with the specified key in the dictionary.
+    /// </summary>
+    /// <typeparam name="T">The type of the value to retrieve.</typeparam>
+    /// <param name="key">The key whose value to get.</param>
+    /// <param name="value">When this method returns, contains the value associated with the specified key,
+    /// if the key is found; otherwise, the default value for the type T.</param>
+    /// <returns>True if the key is found and the value is of type T, false otherwise.</returns>
+    public bool TryGetValue<T>(TKey key, out T value)
+    {
+        if (dictionary.TryGetValue(key, out var objValue) && objValue is T)
```
Comment (TonyTroeff) [inline TryAtSoftware.Extensions.Collections/FluidDictionary.cs:42]: I think you should be able to utilize pattern matching at max and avoid the unnecessary cast on line 44.

File: TryAtSoftware.Extensions.Collections/FluidDictionary.cs [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    /// <exception cref="KeyNotFoundException">Thrown when the key is not found in the dictionary.</exception>
+    /// <exception cref="InvalidCastException">Thrown when the value associated with the key is not of type T.</exception>
+    public T GetRequiredValue<T>(TKey key)
+    {
+        if (!dictionary.TryGetValue(key, out var objValue))
+        {
+            throw new KeyNotFoundException($"Key '{key}' not found in the dictionary.");
+        }
+
+        if (!(objValue is T))
```
Comment (TonyTroeff) [inline TryAtSoftware.Extensions.Collections/FluidDictionary.cs:88]: This looks unpleasant to me. Let's use `is not`.

File: TryAtSoftware.Extensions.Collections/FluidDictionary.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    public IEnumerable<TKey> Keys => dictionary.Keys;
+
+    /// <summary>
+    /// Removes the value associated with the specified key from the dictionary.
+    /// </summary>
+    /// <param name="key">The key whose value to remove.</param>
+    /// <returns>True if the value is successfully removed; otherwise, false.</returns>
+    public bool Remove(TKey key)
+    {
+        if (key == null)
```
Comment (TonyTroeff) [inline TryAtSoftware.Extensions.Collections/FluidDictionary.cs:108]: We recommend using `is` for null checks.

**Review by TonyTroeff — CHANGES_REQUESTED** [review]
My initial review offers some suggestions for code quality improvements.

After they are applied, I will review the PR one more time and will focus on CX (consumer experience) - whether the documentation is understandable, if methods and classes are named and behave as expected.

**Comment by TonyTroeff** [comment]
I just saw that Sonar does not work OOTB for forked repositories. I will try to address this issue as soon as possible.

**Comment by postmeback** [comment]
Thanks for the suggestion. I will update the PR.

File: TryAtSoftware.Extensions.Collections/FluidDictionary.cs (hunk 5 above)
Comment (postmeback) [reply TryAtSoftware.Extensions.Collections/FluidDictionary.cs]: Done

File: TryAtSoftware.Extensions.Collections/FluidDictionary.cs (hunk 4 above)
Comment (postmeback) [reply TryAtSoftware.Extensions.Collections/FluidDictionary.cs]: Implemented accordingly

**Comment by postmeback** [comment]
PR is updated as suggested. Requesting for review.

**Comment by TonyTroeff** [comment]
The workflow process for contributors is fixed. Can you synchronize your fork with the latest changes?

**Comment by postmeback** [comment]
Hi @TonyTroeff , any update ?

File: TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs (hunk 1 above)
Comment (TonyTroeff) [reply TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs:10]: This comment does not seem fixed to me.

File: TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj [hunk 6]
```diff
@@ -25,6 +25,7 @@
 
     <ItemGroup>
       <ProjectReference Include="..\TryAtSoftware.Extensions.Collections\TryAtSoftware.Extensions.Collections.csproj" />
+      <ProjectReference Include="..\TryAtSoftware.Extensions.Reflection.Tests\TryAtSoftware.Extensions.Reflection.Tests.csproj" />
```
Comment (TonyTroeff) [inline TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj:28]: We do not need to cross-reference test projects.

**Review by TonyTroeff — CHANGES_REQUESTED** [review]
There are a few things that must be done before going on with the next stage of review:
- The new tests are failing in the pipeline.
- When they are fixed, SonarCloud will analyze the new code and all code smells, bugs, etc. must be fixed.

**Comment by postmeback** [comment]
image


Test rectified

File: TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs (hunk 1 above)
Comment (postmeback) [reply TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs:10]: In order to get the random number, I have used `using TryAtSoftware.Extensions.Reflection.Tests.Randomization;`

File: TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj (hunk 6 above)
Comment (postmeback) [reply TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj:28]: In order to get the random number, it has been done.

Comment (TonyTroeff) [reply TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj:28]: Use `TryAtSoftware.Randomizer` instead.

Comment (postmeback) [reply TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj:28]: I am unable to figure out, how to use `Randomizer`, requesting for your help
Meanwhile, I have removed cross-referencing and commented the first test function.

Comment (TonyTroeff) [reply TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj]: You can check out this PR for a better understanding: https://github.com/TryAtSoftware/Extensions/pull/66

Comment (postmeback) [reply TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj]: Thanks for it. I have implmented the code changes accordingly.

Comment (postmeback) [reply TryAtSoftware.Extensions.Collections.Tests/TryAtSoftware.Extensions.Collections.Tests.csproj]: Hey, any update?

File: TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs (hunk 1 above)
Comment (TonyTroeff) [reply TryAtSoftware.Extensions.Collections.Tests/FluidDictionaryTests.cs:10]: This is still open. All constant values that are currently being used should be randomized - this includes the keys and the values used in this and the next tests.

**Review by TonyTroeff — CHANGES_REQUESTED** [review]

**Comment by postmeback** [comment]
As my focus of learning has moved, I will not be contributing to this repo. Anyone interested may continue on this.  👍🏾 

Thanks @TonyTroeff for everything.

**Pull request closed without merging**